clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,506 @@
+
+	resultSummary testResultSummary
+}
+
+type podList []workerPodData
+type workerPodInfo struct {
+	podPairCh    chan uniquePodPair
+	workerPodMap map[string]podList
+}
+
+func (l *podList) Pop() (*workerPodData, error) {
+	if len(*l) == 0 {
+		return nil, fmt.Errorf("empty pod list")
+	}
+	lastElement := (*l)[len(*l)-1]
+	*l = (*l)[:len(*l)-1]
+	return &lastElement, nil
+}
+
+func (npm *networkPerformanceMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if err = npm.validate(config); err != nil {
+			return nil, err
+		}
+		if err = npm.start(config); err != nil {
+			return nil, err
+		}
+	case "gather":
+		summary, err := npm.gather()
+		if err != nil && !errors.IsMetricViolationError(err) {
+			return nil, err
+		}
+		klog.Info("metric: ", summary)
+		return []measurement.Summary{summary}, err
+	default:
+		return nil, fmt.Errorf("unknown action")
+	}
+	return nil, nil
+}
+
+func (npm *networkPerformanceMeasurement) start(config *measurement.Config) error {
+	npm.initialize(config)
+
+	if err := npm.prepareCluster(); err != nil {
+		return err
+	}
+
+	if err := npm.createAndWaitForWorkerPods(); err != nil {
+		return err
+	}
+
+	if err := npm.storeWorkerPods(); err != nil {
+		return err
+	}
+
+	if err := npm.initializeInformer(); err != nil {
+		return err
+	}
+
+	if err := npm.executeTest(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (npm *networkPerformanceMeasurement) executeTest() error {
+	switch npm.podRatioType {
+	case oneToOne, manyToMany:
+		npm.execNToMTest()
+	default:
+		return fmt.Errorf("invalid Pod Ratio")
+	}
+	return nil
+}
+
+func (npm *networkPerformanceMeasurement) initialize(config *measurement.Config) {
+	npm.k8sClient = config.ClusterFramework.GetClientSets().GetClient()
+	npm.framework = config.ClusterFramework
+	npm.dynamicClient = config.ClusterFramework.GetDynamicClients().GetClient()
+	npm.podInfo.workerPodMap = make(map[string]podList)
+	npm.podInfo.podPairCh = make(chan uniquePodPair)
+}
+
+func (npm *networkPerformanceMeasurement) initializeInformer() error {
+	gvr, _ := meta.UnsafeGuessKindToResource(crGvk)
+	npm.resourceInterface = npm.dynamicClient.Resource(gvr).Namespace(netperfNamespace)
+	informer, err := getInformer(netperfNamespace, npm.dynamicClient, gvr)
+	if err != nil {
+		return fmt.Errorf("error getting informer:%s", err)
+	}
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+			npm.handleUpdateNetworkTestEvents(newObj)
+		},
+	})
+	npm.stopCh = make(chan struct{})
+	go informer.Run(npm.stopCh)
+	return nil
+}
+
+func (npm *networkPerformanceMeasurement) prepareCluster() error {
+	if err := client.CreateNamespace(npm.k8sClient, netperfNamespace); err != nil {
+		return fmt.Errorf("error while creating namespace: %v", err)
+	}
+
```

Comment:
nit: remove this empty line (same below)

Reply:
Done